core/src/main/scala/chisel3/RawModule.scala

```diff
@@ -130,8 +130,10 @@ abstract class RawModule extends BaseModule {
     case id: DynamicObject => {
       // Force name of the DynamicObject, and set its Property[ClassType] type's ref to the DynamicObject.
       // The type's ref can't be set upon instantiation, because the DynamicObject hasn't been named yet.
+      // This also updates the source Class ref to the DynamicObject ref now that it's named.
       id.forceName(default = "_object", _namespace)
       id.getReference.setRef(id.getRef)
+      id.getSourceClass.foreach(_.setRef(id.getRef, true))
```

Comment:
Since this is called twice and contains some real weirdness, should maybe this functionality be a method on `DynamicObject` that you call? eg.
```scala
id.setSourceClassRef()
```
then `setSourceClassRef` is something like
```scala
def setSourceClassRef(): Unit = this.getSourceClass.foreach(_.setRef(this.getRef, true))
```